File: ios/Sources/MeasureSDK/Swift/Config/ConfigLoader.swift
//
//  ConfigLoader.swift
//  MeasureSDK
//
//  Created by Adwin Ross on 26/08/24.
//

import Foundation

enum ConfigFileConstants {
    static let fileName = "dynamic_config.json"
    static let folderName = "measure"
    static let directory: FileManager.SearchPathDirectory = .cachesDirectory
}

/// A protocol that defines the methods for loading configuration data either a cache or a network source.
protocol ConfigLoader {
    func loadDynamicConfig(onLoaded: @escaping (DynamicConfig) -> Void)
}

/// A base implementation of the `ConfigLoader` protocol.
struct BaseConfigLoader: ConfigLoader {
    private let userDefaultStorage: UserDefaultStorage
    private let fileManager: SystemFileManager
    private let networkClient: NetworkClient
    private let timeProvider: TimeProvider
    private let decoder: JSONDecoder
    private let encoder: JSONEncoder
    private let logger: Logger

    init(userDefaultStorage: UserDefaultStorage,
         fileManager: SystemFileManager,
         networkClient: NetworkClient,
         timeProvider: TimeProvider,
         logger: Logger) {
        self.userDefaultStorage = userDefaultStorage
        self.fileManager = fileManager
        self.networkClient = networkClient
        self.timeProvider = timeProvider
        self.logger = logger

        let decoder = JSONDecoder()
        self.decoder = decoder

        let encoder = JSONEncoder()
        encoder.outputFormatting = [.prettyPrinted]
        self.encoder = encoder
    }

    func loadDynamicConfig(onLoaded: @escaping (DynamicConfig) -> Void) {
        let cachedConfig = loadConfigFromDisk()
        onLoaded(cachedConfig ?? BaseDynamicConfig.default())
        refreshConfigFromServer()
    }

    private func loadConfigFromDisk() -> BaseDynamicConfig? {
        guard let data = fileManager.retrieveFile(
            name: ConfigFileConstants.fileName,
            folderName: ConfigFileConstants.folderName,
            directory: ConfigFileConstants.directory
        ) else {
            return nil
        }

        do {
            return try decoder.decode(BaseDynamicConfig.self, from: data)
        } catch {
            return nil
        }
    }

    private func saveConfigToDisk(_ config: BaseDynamicConfig) {
        do {
            let data = try encoder.encode(config)
            _ = fileManager.saveFile(data: data,
                                     name: ConfigFileConstants.fileName,
                                     folderName: ConfigFileConstants.folderName,
                                     directory: ConfigFileConstants.directory)
        } catch {
            logger.internalLog(level: .error, message: "Failed to save Dynamic config to disk.", error: error, data: nil)
        }
    }

    private func refreshConfigFromServer() {
        let now = timeProvider.now()

        let lastFetch = userDefaultStorage.getConfigFetchTimestamp()
        let cacheControl = userDefaultStorage.getConfigCacheControl()

        let shouldRefresh: Bool

        if lastFetch == 0 || cacheControl == 0 {
            shouldRefresh = true
        } else {
            shouldRefresh = now - lastFetch > cacheControl
        }

        if !shouldRefresh {
            logger.internalLog(
                level: .debug,
                message: "ConfigLoader: CacheControl not expired, skipping refresh",
                error: nil,
                data: nil
            )
            return
        }

        let response = networkClient.getConfig(eTag: userDefaultStorage.getConfigEtag())

        switch response {

        case .success(let config, let eTag, let cacheControl):

            saveConfigToDisk(config)

            userDefaultStorage.setConfigFetchTimestamp(now)
            userDefaultStorage.setConfigCacheControl(Number(cacheControl))

            if let eTag {
                userDefaultStorage.setConfigEtag(eTag)
            }

            logger.internalLog(
                level: .debug,
                message: "ConfigLoader: New config loaded from server successfully",
                error: nil,
                data: nil
            )

        case .notModified:

            userDefaultStorage.setConfigFetchTimestamp(now)

            logger.internalLog(
                level: .debug,
                message: "ConfigLoader: 304 Not Modified",
                error: nil,
                data: nil
            )

        case .error:

            logger.internalLog(
                level: .error,
                message: "ConfigLoader: Failed to load config from server",
                error: nil,
                data: nil
            )
        }
    }
}

File: android/measure/src/main/java/sh/measure/android/config/ConfigLoader.kt
package sh.measure.android.config

import kotlinx.serialization.ExperimentalSerializationApi
import kotlinx.serialization.json.decodeFromStream
import sh.measure.android.executors.MeasureExecutorService
import sh.measure.android.exporter.NetworkClient
import sh.measure.android.logger.LogLevel
import sh.measure.android.logger.Logger
import sh.measure.android.serialization.jsonSerializer
import sh.measure.android.storage.FileStorage
import sh.measure.android.storage.PrefsStorage
import sh.measure.android.tracing.InternalTrace
import sh.measure.android.utils.TimeProvider

internal interface ConfigLoader {
    fun loadDynamicConfig(onLoaded: (DynamicConfig?) -> Unit)
}

internal class ConfigLoaderImpl(
    private val ioExecutor: MeasureExecutorService,
    private val exportExecutor: MeasureExecutorService,
    private val networkClient: NetworkClient,
    private val fileStorage: FileStorage,
    private val prefsStorage: PrefsStorage,
    private val timeProvider: TimeProvider,
    private val logger: Logger,
) : ConfigLoader {
    override fun loadDynamicConfig(onLoaded: (DynamicConfig?) -> Unit) {
        loadConfigFromDisk {
            onLoaded(it)
            refreshConfigFromServer()
        }
    }

    private fun refreshConfigFromServer() {
        exportExecutor.submit {
            val configFile = fileStorage.getConfigFile()
            if (configFile == null) {
                return@submit
            }

            val lastFetchTimestamp = prefsStorage.getConfigFetchTimestamp()
            val cacheControlDuration = prefsStorage.getConfigCacheControl()
            val shouldRefreshConfig = timeProvider.now() - lastFetchTimestamp > cacheControlDuration
            val eTag = prefsStorage.getConfigEtag()

            if (shouldRefreshConfig) {
                val response = networkClient.getConfig(eTag)
                when (response) {
                    is ConfigResponse.Success -> {
                        configFile.writeText(response.body)
                        prefsStorage.setConfigFetchTimestamp(timeProvider.now())
                        response.eTag?.let { prefsStorage.setConfigEtag(it) }
                        prefsStorage.setConfigCacheControl(response.cacheControl)
                        logger.log(
                            LogLevel.Debug,
                            "ConfigLoader: New config loaded from server successfully",
                        )
                    }

                    is ConfigResponse.Error -> {
                        logger.log(
                            LogLevel.Error,
                            "ConfigLoader: Failed to load config from server",
                            response.exception,
                        )
                    }

                    ConfigResponse.NotModified -> {
                        logger.log(LogLevel.Debug, "ConfigLoader: 304 Not Modified")
                    }
                }
            } else {
                logger.log(
                    LogLevel.Debug,
                    "ConfigLoader: CacheControl not expired, skipping refresh",
                )
            }
        }
    }

    @OptIn(ExperimentalSerializationApi::class)
    private fun loadConfigFromDisk(onLoaded: (DynamicConfig?) -> Unit) {
        ioExecutor.submit {
            InternalTrace.trace(
                { "msr-load-config" },
                {
                    val file = fileStorage.getConfigFile()
                    if (file != null && file.exists() && file.length() > 0) {
                        try {
                            val config = file.inputStream().use { stream ->
                                jsonSerializer.decodeFromStream<DynamicConfig?>(stream)
                            }
                            logger.log(
                                LogLevel.Debug,
                                "ConfigLoader: Config loaded successfully from disk",
                            )
                            onLoaded(config)
                        } catch (e: Exception) {
                            logger.log(LogLevel.Error, "ConfigLoader: Failed to load config", e)
                            onLoaded(null)
                        }
                    } else {
                        logger.log(LogLevel.Debug, "ConfigLoader: No config found on disk, falling back to defaults")
                        onLoaded(null)
                    }
                },
            )
        }
    }
}
